pkg/virt-controller/watch/migration.go

```diff
@@ -0,0 +1,717 @@
+		log.Log.Object(migration).Error("target pod disappeared during migration")
+	} else if migration.TargetIsHandedOff() && vmi.Status.MigrationState == nil {
+		migrationCopy.Status.Phase = virtv1.MigrationFailed
+		c.recorder.Eventf(migration, k8sv1.EventTypeWarning, FailedMigrationReason, "VMI's migration state was cleared during the active migration.")
+		log.Log.Object(migration).Error("vmi migration state cleared during migration")
+	} else if migration.TargetIsHandedOff() &&
+		vmi.Status.MigrationState != nil &&
+		vmi.Status.MigrationState.MigrationUID != migration.UID {
+
+		migrationCopy.Status.Phase = virtv1.MigrationFailed
+		c.recorder.Eventf(migration, k8sv1.EventTypeWarning, FailedMigrationReason, "VMI's migration state was taken over by another migration job during active migration.")
+		log.Log.Object(migration).Error("vmi's migration state was taken over by another migration object")
+	} else if vmi.Status.MigrationState != nil &&
+		vmi.Status.MigrationState.MigrationUID == migration.UID &&
+		vmi.Status.MigrationState.Failed {
+
+		migrationCopy.Status.Phase = virtv1.MigrationFailed
+		c.recorder.Eventf(migration, k8sv1.EventTypeWarning, FailedMigrationReason, "Source node reported migration failed")
+		log.Log.Object(migration).Error("VMI reported migration failed.")
+	} else {
+
+		switch migration.Status.Phase {
+		case virtv1.MigrationPhaseUnset:
+			canMigrate, err := c.canMigrateVMI(migration, vmi)
+			if err != nil {
+				return err
+			}
+
+			if canMigrate {
+				migrationCopy.Status.Phase = virtv1.MigrationPending
+			} else {
+				// can not migrate because there is an active migration already
+				// in progress for this VMI.
+				migrationCopy.Status.Phase = virtv1.MigrationFailed
+				c.recorder.Eventf(migration, k8sv1.EventTypeWarning, FailedMigrationReason, "VMI is not eligible for migration because another migration job is in progress.")
+				log.Log.Object(migration).Error("Migration object ont eligible for migration because another job is in progress")
+			}
+		case virtv1.MigrationPending:
+			if podExists {
+				migrationCopy.Status.Phase = virtv1.MigrationScheduling
+			}
+		case virtv1.MigrationScheduling:
+			if isPodReady(pod) {
+				migrationCopy.Status.Phase = virtv1.MigrationScheduled
+			}
+		case virtv1.MigrationScheduled:
+			if vmi.Status.MigrationState != nil && vmi.Status.MigrationState.TargetNode != "" {
+				migrationCopy.Status.Phase = virtv1.MigrationPreparingTarget
+			}
+		case virtv1.MigrationPreparingTarget:
+			if vmi.Status.MigrationState.TargetNode != "" && vmi.Status.MigrationState.TargetNodeAddress != "" {
+				migrationCopy.Status.Phase = virtv1.MigrationTargetReady
+			}
+		case virtv1.MigrationTargetReady:
+			if vmi.Status.MigrationState.StartTimestamp != nil {
+				migrationCopy.Status.Phase = virtv1.MigrationRunning
+			}
+		case virtv1.MigrationRunning:
+			if vmi.Status.MigrationState.Completed {
```

Comment:
Do we want someting like `.spec.ttlSecondsAfterFinished` on the migration object? I see it more and more that automatic cleanup happens for job-like objects in Kubernetes. We can of course say that it will be deleted together with the VMI.

Reply:
I think we should delete the migration once the VMI is deleted... So in this controller, if VMI doesn't exist, delete myself.

Reply:
fixed

Reply:
Maybe we can just delete an old migration object the moment we start a new migration? Why do we need to keep old objects? But, yeah, not important at the moment.